modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/util/DDMDataDefinitionConverterImpl.java

```diff
@@ -262,6 +323,75 @@ private void _upgradeLinkToPageField(DDMFormField ddmFormField) {
 		ddmFormField.setType("link_to_layout");
 	}
 
+	private DDMForm _upgradeNestedFields(DDMForm ddmForm) {
+		if (!_hasNestedFields(ddmForm)) {
+			return ddmForm;
+		}
+
+		DDMForm newDDMForm = new DDMForm();
+
+		newDDMForm.setAvailableLocales(ddmForm.getAvailableLocales());
+		newDDMForm.setDefaultLocale(ddmForm.getDefaultLocale());
+		newDDMForm.setDefinitionSchemaVersion(
+			ddmForm.getDefinitionSchemaVersion());
+
+		for (DDMFormField ddmFormField : ddmForm.getDDMFormFields()) {
+			if (ListUtil.isEmpty(ddmFormField.getNestedDDMFormFields())) {
+				newDDMForm.addDDMFormField(ddmFormField);
+
+				continue;
+			}
+
+			DDMFormField fieldSetDDMFormField = _createFieldSetDDMFormField(
+				ddmForm.getDefaultLocale(), ddmFormField.getName() + "FieldSet",
+				ListUtil.fromArray(ddmFormField), ddmFormField.isRepeatable());
+
+			_upgradeNestedFields(
+				ddmFormField.getNestedDDMFormFields(),
+				ddmForm.getDefaultLocale(), newDDMForm, fieldSetDDMFormField);
+
+			fieldSetDDMFormField.setProperty(
+				"rows", _getDDMFormFieldsRows(fieldSetDDMFormField));
+
+			ddmFormField.setNestedDDMFormFields(Collections.emptyList());
+			ddmFormField.setRepeatable(false);
```

Comment:
I don't understand why we should do these actions in the field if we are returning a new DDM form. Could you explain how these lines affect logic, please? :)

Reply:
I need to do that because I'm reusing the ddmFormField, to add it to the new Form.

```
if (ListUtil.isEmpty(ddmFormField.getNestedDDMFormFields())) {
	newDDMForm.addDDMFormField(ddmFormField);

	continue;
}
```
So if the ddmFormField has nested fields I'm converting it into a fieldset, so the fieldset should the repeatable field and the ddmFormField should not be repeatable and should not have nested fields.